### ISSUE TITLE
Restic does not stop on SIGINT

### ISSUE THREAD

**Issue by vaultah** [restic/restic#1434]
## Output of `restic version`
```
restic 0.7.3 (v0.7.3-237-g06bd606)
compiled with go1.9.2 on linux/amd64
```

## How did you run restic exactly?

```
restic backup --repo b2:my-bucket: my-directory --limit-upload 200 -o b2.connections=1
```

## What backend/server/service did you use?

Backblaze B2.

## Expected behavior

Stop shortly after I press `Ctrl + C`.

## Actual behavior

Restic does not stop (or takes very long time to stop) after I press `Ctrl + C`.

The standard `Interrupt received, cleaning up` message gets printed, but then Restic carries on with the upload. Waiting/pressing `Ctrl + C` repeatedly does not help. The `-o b2.connections=...` setting does not seem to affect anything.

## Steps to reproduce the behaviour

1. Create a repo in a Backblaze B2 bucket.
2. Run `restic backup` to upload your files.
3. Press `Ctrl + C`.

I could not reproduce this with "local" repositories.

## Do you have any idea what may have caused this?

No.

## Do you have an idea how to solve the issue?

No.

**Comment by armhold** [comment]
I have noticed this for a while too. See [this comment](https://github.com/restic/restic/issues/1188#issuecomment-325643132). 

FWIW, I'm able to reproduce this with values of 10 and 1 for `b2.connections`. 

In the logs, I see the following immediately when I issue an interrupt:

```
2017/11/21 10:00:16 restic/cleanup.go:73	main.CleanupHandler	19	signal interrupt received, cleaning up
2017/11/21 10:00:16 restic/lock.go:114	main.unlockAll	19	unlocking 1 locks
2017/11/21 10:00:16 cache/backend.go:38	cache.(*Backend).Remove	19	cache Remove(<lock/ed4345817d>)
2017/11/21 10:00:16 b2/b2.go:292	b2.(*b2Backend).Remove	19	Remove <lock/ed4345817d>
```

But the backup continues for almost 10 minutes more before it stops (it does stop and exit eventually). 

```
2017/11/21 10:09:50 restic/lock.go:120	main.unlockAll	19	successfully removed lock
```

I noticed a few things. 

1. Since the lock cleanup code is the first to insert itself via AddCleanupHandler(), it runs before the other cleanup handlers. This means that the backend lock file could potentially get removed while the backup is still running. I think that's a problem in itself.

2. The lock cleanup code is blocking on be.sem.GetToken(). That's as it should be, but it will also block the other cleanup handlers. One very important cleanup handler is in global.go:

```
	globalOptions.ctx, cancel = context.WithCancel(context.Background())
	debug.Log("global: AddCleanupHandler")
	AddCleanupHandler(func() error {
		cancel()
		return nil
	})
```

If I understand the code correctly, that cleanup handler is meant to signal cancellation throughout the codebase. So I think what's happening here is that the lock cleanup is in contention with the backup code for a semaphore token. Eventually the lock cleanup gets lucky and grabs a token, allowing the cancellation function to run, which should cancel the backup goroutine(s).

3. In cmd_backup, LoadIndex() and Archive() are run with context.TODO() rather than globalOptions.ctx. So I'm not sure how cancellation is working *at all*.  I might have misunderstood something, but this seems wrong?

**Comment by armhold** [comment]
> So I'm not sure how cancellation is working at all

Hmm, I think the way it's "working" is that eventually when the lock removal finally gets a token and returns, cancel() finally runs (effectively as a no-op), and then the entire process ends via os.Exit(). So no graceful shutdown really happening.

**Comment by MichaelEischer** [comment]
I took a closer look at the code and the context handling looks like its mostly fixed by now.

The global context is also cancelled before starting to remove the lock files: the init handler in `global.go` is called before the one in `lock.go`. So with the fixes from #2933 this issue should be resolved.

We could make it explicit that the lock cleanup handler must run after `cancel()` though.
